### PR TITLE
Address TDVP Schmitt unexpected nan and tests.

### LINKED ISSUE #1552: Add more reliable tests for TDVP/TDVPSchmitt
As highlighted in #1550 our tests are not good enough.

We should add a more reliable test that catches bugs by integrating an exact ansatz (LogStateVec) on a 2 spin problem and check that the dynamics matches exactly or approximately the exact one.

### PR DESCRIPTION
**Pull request by Daniel-Haas-B**
Following [previous discussions](https://github.com/orgs/netket/discussions/1959), there was an unexpected behaviour with TDVP driver  when the variance of `QEdata` is too close to zero. In such cases, the signal to noise ratio would simply become undefined, and mess up the entire time evolution. 

The driver already has a cut-off ratio`rcond` for small singular $\sigma_k$ so I propose we use it as  

```python
sigma_k = jnp.maximum(jnp.sqrt(stats.var(QEdata, axis=0)), rcond)
```

If my logic makes sense, I believe this might be more elegant than forcefully setting $\dot{\eta}_k = 0$ for those directions where the variance is zero because it still allows the soft regularisation term `snr_atol` to do its job. Another option which I did not implement would be to to do 

```python
sigma_k = jnp.sqrt(stats.var(QEdata, axis=0))
snr = jnp.where(
    jnp.isclose(sigma_k, 0), jnp.zeros_like(jnp.abs(rho)), jnp.abs(rho) * jnp.sqrt(n_samples) / sigma_k
)
```

or even 

```python
sigma_k = jnp.sqrt(stats.var(QEdata, axis=0))
snr = jnp.abs(rho) * jnp.sqrt(n_samples) / sigma_k
snr = jnp.nan_to_num(snr, nan=0.0)
```

Note that
```python
sigma_k = jnp.sqrt(stats.var(QEdata, axis=0))
snr = jnp.where(
    jnp.isclose(sigma_k_old, 0), jnp.zeros_like(rho), jnp.abs(rho) * jnp.sqrt(n_samples) / sigma_k
)
```
alone does not work because `rho` is complex leading to `snr_atol / snr: [ inf+nanj ...]`.

Importantly, the tests here contained also closes #1552, and would at least have caught the error that motivated that issue. I must add also that this test takes around 20 seconds in my machine.

### CODE REVIEW

File: netket/experimental/driver/tdvp_schmitt.py [hunk 1]
```diff
@@ -225,8 +225,9 @@ def _impl(parameters, n_samples, E_loc, S, rhs_coeff, rcond, rcond_smooth, snr_a
     QEdata = Q.conj() * ΔE_loc
     rho = V.conj().T @ F
 
-    # Compute the SNR according to Eq. 21
-    snr = jnp.abs(rho) * jnp.sqrt(n_samples) / jnp.sqrt(stats.var(QEdata, axis=0))
+    # Compute the SNR according to Eq. 21 but taking care of where sigma_k is zero
+    sigma_k = jnp.maximum(jnp.sqrt(stats.var(QEdata, axis=0)), rcond)
+    snr = jnp.abs(rho) * jnp.sqrt(n_samples) / sigma_k
```
Comment (PhilipVinc) [inline netket/experimental/driver/tdvp_schmitt.py:230]: I'm not against this approach, but the current documentation states that `rcond` is the (relative) cutoff for small eigenvalues. $$ \sigma_k $$ here is instead something like the variance/uncertainty of one eigenvalue, which might very well be large or small...
while both are cutoffs, is it ok to use the same number in both places? I have some doubts... 

--
My reasoning is the following:

In general, if $$\sigma_k$$ is very small, the SNR is well defined and will be extremely large, meaning that in `regularizer2`, below, we will not be regularising it at all.

I think the problem here is really only for $$\sigma_k=0$$ and $$\rho = 0 $$ , because in that case we get 0/0 which is NaN.  I'm not sure what other situation might yield a NaN in the current code.
If only $$\sigma_k$$ is small, then the snr will be something divided by 0 which is infinity, which is correctly handled in the code later. 

Indeed, if an eigenvalue is very small (0) and the variance is very small (0) it's impossible to define what is the SNR (there is no signal, but no noise either).
What should we be doing in that case? As there is no signal, we should not be moving.

To recap: I think the current code is correct for even 0 snr, and is only 'broken' when snr=0 and $$\rho=0$$.
In this case, I believe the right thing to do would be to set `eta_p=0` (not move) explicitly...

What do you think?

File: test/dynamics/test_driver.py [hunk 2]
```diff
@@ -315,6 +317,88 @@ def norm(x):
...
+        psi_t = U @ initial_state
+        for name, op_matrix in obs_matrices.items():
+            exp_t = np.vdot(psi_t, op_matrix @ psi_t).real
+            expectations[name].append(exp_t)
+
+    return expectations
+
+
+def test_tdvp_drivers():
+    """Test time evolution comparing TDVP methods against exact evolution for a mean-field"""
```
Comment (PhilipVinc) [inline test/dynamics/test_driver.py:340]: Can you add a comment to this issue and the previous discussion? something liek
```
# This test verifies a case where SNR = Rho = 0 which used to give NaNs in TDVP Schmitt but not standard TDVP.
# See bug report https://github.com/orgs/netket/discussions/1959 and PR to fix it
# https://github.com/netket/netket/pull/1960
```

File: netket/experimental/driver/tdvp_schmitt.py (hunk 1 above)
Comment (Daniel-Haas-B) [reply netket/experimental/driver/tdvp_schmitt.py:230]: Ok, I completely agree. Do we also agree that this taken care by
```python3
    sigma_k_is_zero = jnp.isclose(sigma_k, 0, atol=1e-15)
    rho_is_zero = jnp.isclose(rho, 0, atol=1e-15)
    snr = jnp.where(
        sigma_k_is_zero & rho_is_zero,
        0.0,
        jnp.where(
            sigma_k_is_zero,
            jnp.inf,
            jnp.abs(rho) * jnp.sqrt(n_samples) / sigma_k,
        )
    )

```
? I just want to avoid the code carries out the NaNs just to filter them only at `eta_p`. This way, `eta_p` will be zero due to the `rho` multiplication at `eta_p = ev_inv * regularizer2 * rhs_coeff * rho, and the NaNs are avoided from the start.

Comment (PhilipVinc) [reply netket/experimental/driver/tdvp_schmitt.py:230]: Is there a particular reason you want to avoid nan being generated?

Regardless, in the code you wrote above you could drop the `jnp.where` in
```python
        jnp.where(
            sigma_k_is_zero,
            jnp.inf,
            jnp.abs(rho) * jnp.sqrt(n_samples) / sigma_k,
        )
```

because you'd get infinity/very large number anyway?

```python
In [17]: sigma_k=jnp.array([0.0, 1e-16])
In [18]: sigma_k_is_zero = jnp.isclose(sigma_k, 0, atol=1e-15)
In [19]: 1/sigma_k
Out[19]: Array([   inf, 1.e+16], dtype=float32)
```
what are you trying to avoid?

**Comment by PhilipVinc** [comment]
bump @Daniel-Haas-B , any update on this?

**Review by PhilipVinc — APPROVED** [review]
I've implemented the last changes I had suggested, and I am merging this if test pass.

Thanks for the work and debugging the issue!
